### PR TITLE
Fix a DST error in date_histogram

### LINKED ISSUE #50265: date_histogram parse_exception in time_zone America/Phoenix
**Elasticsearch version** (`bin/elasticsearch --version`):  8.0.0-SNAPSHOT

**Plugins installed**: []

**JVM version** (`java -version`): 13.0.1

**OS version** (`uname -a` if on a Unix-like system):  Darwin Kernel Version 18.7.0

**Description of the problem including expected versus actual behavior**:

A `date_histogram` with a `fixed_interval` range  of  `1ms` to `999ms` in the timezone `America/Phoenix` causes a "parse_exception". In other timezones e.g. `Europe/Berlin` it works

**Steps to reproduce**:

 1. Try the following request when you've got our Kibana sample flight test data installed
```
GET kibana_sample_data_flights/_search
{
  "aggs": {
    "2": {
      "date_histogram": {
        "field": "timestamp",
        "fixed_interval": "1ms",
        "time_zone": "America/Phoenix",
        "min_doc_count": 1
      }
    }
  }
}
```
Here's the error message of the result:
```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "parse_exception",
        "reason" : "failed to parse date field [-68659199999] with format [epoch_millis]: [failed to parse date field [-68659199999] with format [epoch_millis]]"
      }
    ],
    "type" : "search_phase_execution_exception",
    "reason" : "all shards failed",
    "phase" : "query",
    "grouped" : true,
    "failed_shards" : [
      {
        "shard" : 0,
        "index" : "kibana_sample_data_flights",
        "node" : "N2J6g-DxSySnU5YiP-9aAA",
        "reason" : {
          "type" : "parse_exception",
          "reason" : "failed to parse date field [-68659199999] with format [epoch_millis]: [failed to parse date field [-68659199999] with format [epoch_millis]]",
          "caused_by" : {
            "type" : "illegal_argument_exception",
            "reason" : "failed to parse date field [-68659199999] with format [epoch_millis]",
            "caused_by" : {
              "type" : "date_time_parse_exception",
              "reason" : "Failed to parse with all enclosed parsers"
            }
          }
        }
      }
    ]
  },
  "status" : 400
}
```

**Provide logs (if relevant)**:
```
   │ info [o.e.a.s.TransportSearchAction] [kertal-elastic.local] All shards failed for phase: [query]
   │      org.elasticsearch.ElasticsearchParseException: failed to parse date field [-68659199001] with format [epoch_millis]: [failed to parse date field [-68659199001] with format [epoch_millis]]
   │      	at org.elasticsearch.common.time.JavaDateMathParser.parseDateTime(JavaDateMathParser.java:233) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.common.time.JavaDateMathParser.parse(JavaDateMathParser.java:74) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.index.mapper.DateFieldMapper$DateFieldType.parseToLong(DateFieldMapper.java:401) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.index.mapper.DateFieldMapper$DateFieldType.isFieldWithinQuery(DateFieldMapper.java:415) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.search.aggregations.bucket.histogram.DateHistogramAggregationBuilder.rewriteTimeZone(DateHistogramAggregationBuilder.java:473) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.search.aggregations.bucket.histogram.DateHistogramAggregationBuilder.innerBuild(DateHistogramAggregationBuilder.java:493) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.search.aggregations.support.ValuesSourceAggregationBuilder.doBuild(ValuesSourceAggregationBuilder.java:313) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.search.aggregations.support.ValuesSourceAggregationBuilder.doBuild(ValuesSourceAggregationBuilder.java:37) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.search.aggregations.AbstractAggregationBuilder.build(AbstractAggregationBuilder.java:139) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.search.aggregations.AggregatorFactories$Builder.build(AggregatorFactories.java:333) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.search.SearchService.parseSource(SearchService.java:790) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.search.SearchService.createContext(SearchService.java:586) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:545) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:348) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.search.SearchService.lambda$executeQueryPhase$1(SearchService.java:340) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.action.ActionListener.lambda$map$2(ActionListener.java:146) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:63) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:58) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:73) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elastic
   │ info search-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:44) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:769) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
   │      	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
   │      	at java.lang.Thread.run(Thread.java:830) [?:?]
   │      Caused by: java.lang.IllegalArgumentException: failed to parse date field [-68659199001] with format [epoch_millis]
   │      	at org.elasticsearch.common.time.JavaDateFormatter.parse(JavaDateFormatter.java:169) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.common.time.JavaDateMathParser.parseDateTime(JavaDateMathParser.java:223) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	... 25 more
   │      Caused by: java.time.format.DateTimeParseException: Failed to parse with all enclosed parsers
   │      	at org.elasticsearch.common.time.JavaDateFormatter.doParse(JavaDateFormatter.java:196) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.common.time.JavaDateFormatter.parse(JavaDateFormatter.java:167) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	at org.elasticsearch.common.time.JavaDateMathParser.parseDateTime(JavaDateMathParser.java:223) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
   │      	... 25 more
```

### PR DESCRIPTION
**Pull request by nik9000**
When `date_histogram` attempts to optimize itself it for a particular
time zone it checks to see if the entire shard is within the same
"transition". Most time zone transition once every size months or
thereabouts so the optimization can usually kicks in.

*But* it crashes when you attempt feed it a time zone who's last DST
transition was before epoch. The reason for this is a little twisted:
before this patch it'd find the next and previous transitions in
milliseconds since epoch. Then it'd cast them to `Long`s and pass them
into the `DateFieldType` to check if the shard's contents were within
the range. The trouble is they are then converted to `String`s which are
*then* parsed back to `Instant`s which are then convertd to `long`s. And
the parser doesn't like most negative numbers. And everything before
epoch is negative.

This change removes the
`long` -> `Long` -> `String` -> `Instant` -> `long` chain in favor of
passing the `long`s directly into the `DateFieldType`. Sadly, this means
that `MappedFieldType` gets a `long` specialized implementation of
`isFieldWithinQuery` which is a little lame, but given that we always
call it with a long in this context it feels like the lesser of two
evils. Not because it is more efficient. It is a little. No. This is the
less evil way to solve this problem because it is less confusing to
trace the execution. The parsing code is fairly complex and we can just
skip it entirely because we already *have* longs.

Closes #50265

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java [hunk 1]
```diff
@@ -419,10 +419,27 @@ public IntervalsSource intervals(String query, int max_gaps, boolean ordered,
...
+    /**
+     * Return whether all values of the given {@link IndexReader} are within the range,
+     * outside the range or cross the range. The default implementation returns
+     * {@link Relation#INTERSECTS}, which is always fine to return when there is
+     * no way to check whether values are actually within bounds.
+     *
+     * Use this instead of
+     * {@link #isFieldWithinQuery(IndexReader, Object, Object, boolean, boolean, ZoneId, DateMathParser, QueryRewriteContext)}
+     * when you *know* the {@code fromInclusive} and {@code toInclusive} are always
+     * floats.
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java]: I think you mean they're always longs?

**Review by not-napoleon — APPROVED** [review]
LGTM.  Nice work finding this one.  I like the solution, not going through some weird coercion chain seems like a win.

**Comment by nik9000** [comment]
> LGTM. Nice work finding this one. I like the solution, not going through some weird coercion chain seems like a win.

@jimczi actually found it! He just didn't get a chance to fix it before going on a long trip. He did mention that @jpountz may want to look at it so I added him as a reviewer too.

File: server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java (hunk 1 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java]: +1

**Review by jpountz — COMMENTED** [review]
I'd really like to avoid adding the method to MappedFieldType, can we have it only live in DateFieldType and do an instanceof check in the aggregator, and disable the optimization if the instanceof check fails? I think we also need javadocs to clarify that the longs are a number of millis, which is not obvious at all if the field is a `date_millis`. By the way we might want to have an explicit test for `date_millis` to make sure it does the right thing with this new method?

**Comment by nik9000** [comment]
> I'd really like to avoid adding the method to MappedFieldType, can we have it only live in DateFieldType and do an instanceof check in the aggregator, and disable the optimization if the instanceof check fails? I think we also need javadocs to clarify that the longs are a number of millis, which is not obvious at all if the field is a `date_millis`. By the way we might want to have an explicit test for `date_millis` to make sure it does the right thing with this new method?

I can do all those things. I'm not a fan of `instanceof` for this sort of thing but if you feel like it is better I'm happy to do it. I agree having a method that only makes sense for dates on `MappedFieldType` is a bit strange.

**Comment by nik9000** [comment]
Good call on the `date_nanos` test, and the resolution of the interface.

**Comment by nik9000** [comment]
@jpountz, I've added something to properly handle nanos. It feels properly paranoid to me now.

**Comment by nik9000** [comment]
@elasticmachine run  elasticsearch-ci/packaging-sample-matrix-unix

**Review by jpountz — APPROVED** [review]
Thanks Nik, this looks good to me now. As you guessed, I meant `date_nanos`, not `date_millis`. :) I also like the way you refactored the rewrite of the time zone, I find it easier to read now.

**Comment by nik9000** [comment]
> I also like the way you refactored the rewrite of the time zone, I find it easier to read now.

**Comment by nik9000** [comment]
Thanks @not-napoleon and @jpountz !

**Pull request merged**